documentation/vespa-plugins.html

```diff
@@ -20,55 +20,94 @@
-FeedAndSearchSystemTest.java</a>.
-A system test is an integration test that tests the application API endpoints.
-</p>
+<table class="table">
+<tr id="build">
+  <th>Build</th>
+  <td>
+    <p>All the Vespa sample applications use the <a href="bundle-plugin.html">bundle plugin</a>
+    to build the Components.</p>
+  <td>
+</tr>
+<tr id="configure">
+  <th>Configure</th>
+  <td>
+    <p>
+    A key Vespa feature is code and configuration concistency,
+    deployed using an <a href="cloudconfig/application-packages.html">application package</a>.
+    This ensures that code and configuration is in sync, and loaded atomically when deployed.
+    This is done by generating config classes from config definition files.
+    In Vespa and application code,
+    configuration is hence accessed through generated config classes -
+    read more in the <a href="cloudconfig/config-introduction.html">
+    Cloud Configuration System</a>.
+    </p><p>
+    The Maven target <em>generate-sources</em> (invoked by <em>mvn install</em>) uses
+    <code><a href="https://github.com/vespa-engine/sample-apps/blob/master/vespa-cloud/album-recommendation-searcher/src/main/resources/configdefinitions/metal-names.def">
+    metal-names.def</a></code>
+    to generate <code>target/generated-sources/vespa-configgen-plugin/com/mydomain/example/MetalNamesConfig.java</code>.
+    </p><p>
+    After generating config classes, they will resolve in tools like
+    <a href="http://www.jetbrains.com/idea/download/">IntelliJ IDEA</a>.
+    </p>
+  <td>
+</tr>
+<tr id="tests">
+  <th>Tests</th>
+  <td>
+    <p>
+    Examples unit tests are found in
+    <a href="https://github.com/vespa-engine/sample-apps/blob/master/vespa-cloud/album-recommendation-searcher/src/test/java/com/mydomain/example/MetalSearcherTest.java">
+    MetalSearcherTest.java</a>.
+    <code>testAddedOrTerm1</code> and <code>testAddedOrTerm2</code> illustrates two ways of doing the same test:
+    <ul>
+      <li>The first setting up the minimal search chain for <a href="query-language.html">YQL</a> programmatically</li>
+      <li>The second uses <code>
+    <a href="http://javadoc.io/page/com.yahoo.vespa/application/latest/com/yahoo/application/Application.html">
+    com.yahoo.application.Application</a></code>, which sets up the application package and simplifies testing</li>
+    </ul>
+    Read more in <a href="testing.html">testing</a>.
+    </p>
+  <td>
+</tr>
+<tr id="automate">
+  <th>Automate</th>
+  <td>
+    <p>
+    Add testing to automate deployments:
+    <ul>
+      <li><em>System Tests</em> for functional tests -
+        a system test is an integration test that tests the application API endpoints</li>
+      <li><em>Staging Tests</em> to validate the instance before and after the change,
+          i.e. two test runs with potentially different tests</li>
```

Comment:
Perhaps something to further distinguish this from the above — staging tests validate that state in the cluster is not broken by an upgrade of the Vespa runtime, or application code or config?

Reply:
maybe. It is not _that_ interesting for the user that we also test during vespa upgrades, in the context of this document. can add some more details on code/config